packages/candid/src/idl.ts

```diff
@@ -924,15 +924,29 @@ export class RecordClass extends ConstructType<Record<string, any>> {
     const x: Record<string, any> = {};
     let idx = 0;
     for (const [hash, type] of record._fields) {
+      let [expectKey, expectType] = this._fields[idx];
+
+      // skip expected optional fields that are not present in the data
+      while (
+        (expectType instanceof OptClass || expectType instanceof ReservedClass) &&
+        idx < this._fields.length &&
+        idlLabelToId(this._fields[idx][0]) !== idlLabelToId(hash)
+      ) {
+        x[expectKey] = [];
+        idx++;
+        [expectKey, expectType] = this._fields[idx];
+      }
+
       if (idx >= this._fields.length || idlLabelToId(this._fields[idx][0]) !== idlLabelToId(hash)) {
```

Comment:
Unfortunately not because then we would go to the next pair of `const [hash, type] of record._fields` without having processed the current one. I have now switched the whole thing to a `while` loop.